backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitChannelPoolImpl.java

```diff
@@ -30,24 +32,37 @@
 import org.apache.commons.pool2.impl.DefaultPooledObject;
 import org.apache.commons.pool2.impl.GenericObjectPool;
 import org.apache.james.util.MemoizedSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.github.fge.lambdas.Throwing;
 import com.rabbitmq.client.Channel;
 import com.rabbitmq.client.Connection;
 
 public class RabbitChannelPoolImpl implements RabbitMQChannelPool {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(RabbitChannelPoolImpl.class);
+
     private static class ChannelBasePooledObjectFactory extends BasePooledObjectFactory<Channel> {
-        private final Supplier<Connection> connection;
+        private final Supplier<Connection> rabbitConnection;
+        private final AtomicBoolean connectionOpened;
 
         public ChannelBasePooledObjectFactory(RabbitMQConnectionFactory factory) {
-            this.connection = MemoizedSupplier.of(
-                    Throwing.supplier(() -> factory.create()).sneakyThrow());
+            this.connectionOpened = new AtomicBoolean();
+            this.rabbitConnection = MemoizedSupplier.of(
```

Comment:
Maybe this 'is inited' logic deserve to be in memoizedSupplier?

So that we can just do a `supplier.ifInitlialized(Connection::close);`

That would avoid the complexity of tracking the supplier state in this class.

What would you think of it?

Reply:
And you could use an AtomicReference instead of an AtomicBoolean to make the code simpler

Reply:
I eliminated `AtomicBoolean` and move it inside the new `MemorizedSupplier`. Please review them(5bcf787, 807152b) or below commits

Reply:
Well, did you tried to use AtomicReference ? I really think code would be simpler